File: s32/bench_cpu.py
"""bench_cpu.py - meta' del confronto Python/LuaJIT (vedi bench_cpu.lua
per l'altra meta'): carica l'immagine di memoria prodotta da
dump_cpu_state.py e cronometra N chiamate a cpu.run() con cpu.py
(Python/Cython, quello che gira davvero nel motore) - stesso identico
lavoro che bench_cpu.lua fa con cpu.lua, per un confronto onesto a
parita' di input.

Uso:
    python3 bench_cpu.py [cpu_bench_mem.bin] [start_pc esadecimale] [n_frame]

Esempio:
    python3 bench_cpu.py cpu_bench_mem.bin 0x1000 1000
"""
import sys
import os
import time

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from cpu import CPU

mem_path = sys.argv[1] if len(sys.argv) > 1 else "cpu_bench_mem.bin"
start_pc = int(sys.argv[2], 0) if len(sys.argv) > 2 else 0x001000
n_frames = int(sys.argv[3]) if len(sys.argv) > 3 else 1000

with open(mem_path, "rb") as f:
    data = f.read()

cpu = CPU()
cpu.mem[:] = data

cpu_total = 0.0
total_instr = 0
for _ in range(n_frames):
    t0 = time.perf_counter()
    n = cpu.run(start_pc, input_byte=0)
    t1 = time.perf_counter()
    cpu_total += (t1 - t0)
    total_instr += n

cpu_ms = cpu_total / n_frames * 1000
avg_instr = total_instr / n_frames
us_per_instr = (cpu_total / total_instr) * 1e6 if total_instr else 0.0

print(f"Python (cpu.py: {CPU.__module__}, verifica .so/.py con 'import cpu; print(cpu.__file__)')")
print(f"  {n_frames} frame, cpu totale {cpu_total*1000:.2f}ms, "
      f"media {cpu_ms:.4f}ms/frame, istruzioni media {avg_instr:.1f}/frame, "
      f"{us_per_instr:.2f}us/istruzione")

File: s32/bench_cpu.lua
--[[
bench_cpu.lua - meta' del confronto Python/LuaJIT (vedi bench_cpu.py
per l'altra meta'): carica la STESSA immagine di memoria prodotta da
dump_cpu_state.py e cronometra N chiamate a cpu:run() con cpu.lua.

Uso (da lanciare con luajit, non lua "normale" - vedi cpu.lua):
    luajit bench_cpu.lua [cpu_bench_mem.bin] [start_pc] [n_frame]

Esempio:
    luajit bench_cpu.lua cpu_bench_mem.bin 0x1000 1000

Deve girare dalla stessa cartella di cpu.lua (require("cpu") lo cerca
li' di default).
]]
local ffi = require("ffi")

ffi.cdef[[
typedef struct { long tv_sec; long tv_nsec; } timespec_t;
int clock_gettime(int clk_id, timespec_t *tp);
]]
local CLOCK_MONOTONIC = 1
local function now()
    local ts = ffi.new("timespec_t")
    ffi.C.clock_gettime(CLOCK_MONOTONIC, ts)
    return tonumber(ts.tv_sec) + tonumber(ts.tv_nsec) / 1e9
end

local cpu_module = require("cpu")

local mem_path = arg[1] or "cpu_bench_mem.bin"
local start_pc = arg[2] and tonumber(arg[2]) or 0x001000
local n_frames = arg[3] and tonumber(arg[3]) or 1000

local cpu = cpu_module.new()
cpu:load_memory_image(mem_path)

local cpu_total = 0
local total_instr = 0
for i = 1, n_frames do
    local t0 = now()
    local n = cpu:run(start_pc, 0)
    local t1 = now()
    cpu_total = cpu_total + (t1 - t0)
    total_instr = total_instr + n
end

local cpu_ms = cpu_total / n_frames * 1000
local avg_instr = total_instr / n_frames
local us_per_instr = total_instr > 0 and (cpu_total / total_instr) * 1e6 or 0

print("LuaJIT (cpu.lua)")
print(string.format("  %d frame, cpu totale %.2fms, media %.4fms/frame, istruzioni media %.1f/frame, %.2fus/istruzione",
    n_frames, cpu_total * 1000, cpu_ms, avg_instr, us_per_instr))
